### PR TITLE
[Jobs] Add template for how metadata could be passed

### PR DESCRIPTION
**Pull request by peterghaddad**
## Why are these changes needed?

Enables users to pass metadata for CLI calls that use the `--address` flag.

## Related issue number

#27748

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peterghaddad** [comment]
Hi @jjyao, i'm having a hard time wrapping my head around adding the headers to the gRPC stubs. I am aware that for the client we utilize a specific DataClass but the CLI doesn't go down this same trace. Any comments would be helpful.

**Comment by jjyao** [comment]
Thanks for the contribution. I'll take a look soon!

**Comment by jjyao** [comment]
https://github.com/ray-project/ray/issues/27748#issuecomment-1249473449

**Comment by peterghaddad** [comment]
@jjyao Could we review this?

File: python/ray/_private/utils.py
Comment (jjyao) [inline python/ray/_private/utils.py:1263]: based on the ray client implementation, the metadata needs to be passed to the stub method call? Does that mean we need to find all the stub calls and add metadata?

File: python/ray/_private/utils.py
Comment (jjyao) [inline python/ray/_private/utils.py:1261]: I have one high level question:

Is this needed only when you want to run the CLIs remotely instead of within the cluster. Like what if you can ssh to the head node and run there, do you still need the custom metadata?

**Review by jjyao — COMMENTED** [review]
Really sorry for the late review.

File: python/ray/_private/utils.py
Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: Well, only the stubs that are utilized by the CLI when accessing the Ray Cluster. This was implemented with ray.init() meaning users can pass `ray.init(_metadata={})`.

File: python/ray/_private/utils.py
Comment (peterghaddad) [reply python/ray/_private/utils.py:1261]: Yes that is correct, but only admins can ssh onto the Head Node. We have a reverse proxy in front of Ray which acts as an authentication layer meaning it would have to be passed as a header. There were quite a few companies at Ray Summit that have a similar architecture, so think this would be a helpful feature.

Comment (jjyao) [reply python/ray/_private/utils.py:1261]: Yea, I agree we need to implement it.

File: python/ray/_private/utils.py
Comment (jjyao) [reply python/ray/_private/utils.py:1263]: So what's the next step of this PR? Find all the related stubs and add metadata?

Also I realized that some CLIs (e.g. job submission CLI) are talking to the cluster through HTTP instead of GRPC, we should also think about that as well (but I think it's out of scope of this PR).

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: I think I could use some guidance. Unsure which CLI commands would be best to support this feature, and areas at code to start looking into.

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: For example, which CLI commands should support metadata? Is there a best format for adding new params for cli commands? Pointing to code files would be helpful, just want to make sure we implement this correctly. Thank you!!!

File: python/ray/_private/utils.py
Comment (peterghaddad) [reply python/ray/_private/utils.py:1261]: Hope that clarifies. Let me know if you have any other questions!

File: python/ray/_private/utils.py
Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: Hi @jjyao, wanted to follow up on this discussion. Thanks!

Comment (jjyao) [reply python/ray/_private/utils.py:1263]: Sorry for the late reply. I feel the most important CLIs to support are actually `ray job submit` (https://github.com/ray-project/ray/blob/75b206e419f88d25e5e08777b2548600287e2120/dashboard/modules/job/cli.py#L158), `ray logs`, `ray list`  which are based on https instead of grpc. But I think the same principle applies. Do you mind if we start with those?

Comment (jjyao) [reply python/ray/_private/utils.py:1263]: Or we can do `ray status` which is based on grpc: https://github.com/ray-project/ray/blob/75b206e419f88d25e5e08777b2548600287e2120/python/ray/scripts/scripts.py#L1872

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: I like the path of going both ways. I'm curious if there is an init command that will persist this metadata or cookies that are passed to all calls?

For example, could we use an environment variable?

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: The JobSubmissionClient supports this already, if I am reading the code correctly: https://github.com/ray-project/ray/blob/75b206e419f88d25e5e08777b2548600287e2120/dashboard/modules/dashboard_sdk.py#L125

Can you take a quick poke and verify? Thanks so much for the collaboration!

Comment (jjyao) [reply python/ray/_private/utils.py:1263]: This metadata is not for authentication but it also supports http headers that can be used for authentication. Seems you can pass in headers if you are using job submission apis programatically. But currently you cannot pass those using `ray job submit` CLI. @architkulkarni is my understanding correct? Is this something we want to support in CLI as well. I think @peterghaddad is willing to contribute here.

Comment (architkulkarni) [reply python/ray/_private/utils.py:1263]: It would be great to support these in the Ray Jobs CLI! I think it just hasn't been prioritized yet.

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: Happy to work these. Do you think an environment variable would be sufficient? This may be the way to go so that the end-user doesn't have to type in authentication metadata per cli command?

Comment (architkulkarni) [reply python/ray/_private/utils.py:1263]: Oh @peterghaddad I just realized I was mistaken-- in the Jobs API, the `metadata` field is used for arbitrary user-provided metadata which is eventually just internally stored at https://github.com/ray-project/ray/blob/master/python/ray/job_config.py#L29, so I don't think it works for the kind of authentication uses you have in mind (it never gets read by anything).  

Can what you're doing be accomplished just using the `headers` option?  If this is missing from the jobs CLI, it would be good to add it there.  Having an option to specify it via an environment variable makes sense to me.

Comment (peterghaddad) [reply python/ray/_private/utils.py:1263]: Believe so, so the `headers` option is what's passed through network traffic? i.e cookies etc.

File: python/ray/_private/utils.py
Comment (peterghaddad) [reply python/ray/_private/utils.py:1261]: @architkulkarni Do all cli command use the same gRPC stub? I believe we could wrap the gRPC stub additional auth if the environment variable exists. If so, can you point out the specific stub call that gets called? Is it the one posted above? Appreciate all the discussion, just want to make sure I am understanding everything correctly.

Comment (jjyao) [reply python/ray/_private/utils.py:1261]: Some CLI commands use http like `ray job sumit` instead of grpc.

Comment (peterghaddad) [reply python/ray/_private/utils.py:1261]: Right. But for the ones that do use gRPC, do they use the same stubs?

Comment (jjyao) [reply python/ray/_private/utils.py:1261]: I think the stubs are `internal_kv_*()` (e.g. `internal_kv_get()`, `internal_kv_list()`). I actually take a look of all the CLIs. I think the ones that we want to run remotely and need authentications are mostly http based CLIs. So I think we should start with those first.

File: python/ray/_private/utils.py
Comment (jjyao) [reply python/ray/_private/utils.py:1263]: Yes.

**Comment by peterghaddad** [comment]
Hi @JiahaoYao, this is ready for review. Mind taking a look?

**Comment by DmitriGekhtman** [comment]
> Hi @JiahaoYao, this is ready for review. Mind taking a look? Unsure why it is showing merge conflicts since I am not seeing these locally?

I think you are looking for @jjyao :)

**Comment by DmitriGekhtman** [comment]
> Unsure why it is showing merge conflicts since I am not seeing these locally?

Hmm, have you pulled the upstream master and either rebased the PR branch on top of master or merged master into the PR branch?

**Comment by peterghaddad** [comment]
Hi @jjyao, is it possible to get another review?

**Comment by architkulkarni** [comment]
Looks like this PR only touches the Jobs API, I will review it.

@peterghaddad it looks like there are some potentially relevant CI failures https://buildkite.com/ray-project/oss-ci-build-pr/builds/6771#0184cfbc-2bcf-4c08-a64f-18b74d30cd39/3334-8357, could you take a look?

**Comment by peterghaddad** [comment]
Hi @architkulkarni, yep! We will start with only the job cli since it utilizes http. Pushed a change for the jobs. Will keep an eye on it. I signed off on the most recent commit. Do I need to re-write history on the DCO job?

**Comment by peterghaddad** [comment]
@architkulkarni I'm having a hard time creating running the tests locally that are failing in the CI. Is there a way to get better logs regarding the CLI test? Locally, every test fails with: ```Traceback (most recent call last):
  File "/Users/e356161/ai-factory/github/ray/python/ray/tests/test_cli.py", line 40, in <module>
    import ray
  File "/Users/e356161/ai-factory/github/ray/venv/lib/python3.9/site-packages/ray/__init__.py", line 109, in <module>
    import ray._raylet  # noqa: E402
  File "python/ray/_raylet.pyx", line 142, in init ray._raylet
ImportError: cannot import name DeferSigint```

I have followed the Documentation on [here](https://docs.ray.io/en/latest/ray-contribute), and I am running on a Mac.

**Comment by architkulkarni** [comment]
Hi @peterghaddad , sorry you're running into this. We want to improve the workflow for contributors, so we'd appreciate any feedback you have on the doc page or anything else.

Because the error is coming from a `.pyx` file which is related to Cython, my guess is this can be fixed by building Ray from source. Could you try https://docs.ray.io/en/latest/ray-contribute/development.html#building-ray-python-only, and if that doesn't work, https://docs.ray.io/en/latest/ray-contribute/development.html#preparing-to-build-ray-on-macos?

**Comment by peterghaddad** [comment]
Hi @architkulkarni, i've tried building Ray from source on a couple different machines. Are there other recommendations such as using PyEnv, Poetry, or building within a Docker container?

**Comment by architkulkarni** [comment]
Hi @peterghaddad, sorry to hear that, the only tips I have are to use a fresh `conda` env and to try running `bazel clean --expunge` in case there's some bad state in bazel.  If you post your build error on [discuss.ray.io](http://discuss.ray.io/) I'm sure we can figure out the issue, and I can pull in more experts to help.

**Comment by architkulkarni** [comment]
Ah, one last thing to try, sometimes errors that look like `File "python/ray/_raylet.pyx"` `ImportError: cannot import name ...` can be fixed with `rm -rf python/ray/thirdparty_files/`, followed by `pip install -e . --verbose` in the `python` directory to trigger the rebuild.  This might be worth a shot

**Comment by peterghaddad** [comment]
@architkulkarni can we open this issue back up?

**Comment by architkulkarni** [comment]
Oh weird, reopening

**Comment by peterghaddad** [comment]
@architkulkarni I am running into some problems testing the Job CLI after building Ray from source. Following these docs: https://docs.ray.io/en/latest/ray-contribute/development.html#preparing-to-build-ray-on-macos

Pretty much the job cli command is missing from the cli tool. Are there additional steps that need to be done? I also posted in Slack (in `#ray-contributors`, and waiting for a response. Any help would be appreciated, so I can test this and push it forward.

**Comment by architkulkarni** [comment]
Responded in Slack, were you able to get a test environment running locally?

**Comment by peterghaddad** [comment]
@architkulkarni I was able to build it locally! Had to do with my pyenv environment and some caching behaviors. I was able to run the linter, test, and run everything locally. I am seeing there are still some CI jobs failing, is there any additional steps that need to be taken? I appreciate all the help!

**Comment by architkulkarni** [comment]
Nice! It looks like there may be some relevant test failures https://buildkite.com/ray-project/oss-ci-build-pr/builds/21403#018805c7-928b-47ce-baf6-55145968770b/3233-9371, could you take a look? Let me know if you can't reproduce the failure locally with `pytest -vs test_cli_integration.py` (for instance).

Unfortunately we have some flaky tests which are potentially unrelated to your PR.  You can check against https://flaky-tests.ray.io/ to help determine whether or not it's related to the PR.

File: dashboard/modules/job/cli.py
Comment (architkulkarni) [inline dashboard/modules/job/cli.py:34]: It looks like `RAY_JOB_HEADERS` is only being read in the CLI codepath. However, some users use the SDK directly.  They may expect that if `RAY_JOB_HEADERS` is set directly, it will be picked up by their `JobSubmissionClient.` Can you fix this? Alternatively, we can defer the `RAY_JOB_HEADERS` to a subsequent PR and just focus on the CLI argument `--headers` in this PR.

File: dashboard/modules/job/cli.py
Comment (architkulkarni) [inline dashboard/modules/job/cli.py:33]: Stray log?

File: dashboard/modules/job/cli.py
Comment (architkulkarni) [inline dashboard/modules/job/cli.py:177]:
```suggestion
        "Please follow JSON formatting, e.g. {'key': 'value'}"
```

Comment (architkulkarni) [reply dashboard/modules/job/cli.py:177]: Please also update the other docstrings

File: doc/source/cluster/running-applications/job-submission/quickstart.rst
Comment (architkulkarni) [inline doc/source/cluster/running-applications/job-submission/quickstart.rst:84]: We'd like to keep the quickstart as minimal as possible, is there somewhere else in the docs it makes sense to include this snippet?

**Review by architkulkarni — COMMENTED** [review]
Implementation looks good, please also add tests for the new functionality.

**Comment by architkulkarni** [comment]
cc @akshay-anyscale for awareness

File: dashboard/modules/job/cli.py
Comment (peterghaddad) [reply dashboard/modules/job/cli.py:34]: Let's defer it, and we can use both!

File: dashboard/modules/job/cli.py
Comment (peterghaddad) [reply dashboard/modules/job/cli.py:177]: Of course!

File: doc/source/cluster/running-applications/job-submission/quickstart.rst
Comment (peterghaddad) [reply doc/source/cluster/running-applications/job-submission/quickstart.rst:84]: I was doing some reading through the sections and noticed a file cli.rst. Would it make sense to include it here. I originally put it here because we do something very similar for RAY_ADDRESS in the quickstart file.

**Comment by peterghaddad** [comment]
@architkulkarni you asked if I could run `pytest -vs test_cli_integration.py` locally.

I am running on MacOS and seeing the following errors. I tried the following but seems to not work. https://github.com/ray-project/ray/issues/24920

Below is the error:

```ERROR node.py:605 -- Failed to connect to GCS. Please check `gcs_server.out` for more detail````

I am going to keep troubleshooting, but wanted to let you know incase you had any ideas in mind. I appreciate the feedback, and think we're in getting close.

**Comment by architkulkarni** [comment]
@peterghaddad That's odd, does this happen on the master branch or only on this branch? I agree `export RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER=1` is a good idea (We plan to improve the developer experience here https://github.com/ray-project/ray/pull/35226) 

I'm not sure what to suggest other than rebuilding it.  Does `ray start --head` fail in general, or does it only fail inside this test?

**Comment by peterghaddad** [comment]
Hey @architkulkarni, running `ray start --head` works fine. I have tried rebuilding the source and also building `main` and experience the same problems. For now, I would use the CI to perform tests but it does not pinpoint problems like I would assume it would locally. I appreciate the response!

**Comment by architkulkarni** [comment]
@peterghaddad Another random thought, here's the first failure: https://buildkite.com/ray-project/oss-ci-build-pr/builds/22591#0188311e-4690-485b-a269-39b89f16fcbb/2078-3639

```
def test_address(self, mock_sdk_client):
--
  | >       _job_cli_group_test_address(mock_sdk_client, "list")
[...]
  | E           AssertionError: Expected call: JobSubmissionClient('arg_addr', False, verify=True)
  | E           Actual call: JobSubmissionClient()
```

I wonder if this failure is somehow causing cascading failures that are harder to debug.  Maybe you can run the individual test like so:
`pytest -vs test_cli.py::TestList::test_address`

**Comment by architkulkarni** [comment]
I don't know how to debug the `Failed to connect to GCS` error unfortunately, other than to ensure there is no currently running Ray cluster before running the tests locally, for example by calling `ray stop` before running the test

**Comment by peterghaddad** [comment]
Hey @architkulkarni, I've built the source with a few different systems and was wondering if there is a way to build source within a container so that I can run the unit tests locally?

**Comment by architkulkarni** [comment]
Hi @peterghaddad, sorry for all the trouble. I'm not an expert on this part, but I don't see any reason why Ray couldn't be built in a container.  I think you'll have a better response on the Ray Slack though where we can pull in some people who might know more. 

As a somewhat extreme workaround, you can try creating a new test file that only has the tests and fixtures you're interested in. And a random thought, is `RAY_ADDRESS` set locally on your machine? If so it might need to be deleted so it doesn't interfere with the tests.

**Comment by peterghaddad** [comment]
@architkulkarni Thanks for the response! Looks like using `master` on Ray no longer produces that GCS connect error with Ray 2.5.0

Are these the results you would expect when running the tests on Master? I am running `pytest test_cli_integration.py` and checked https://flaky-tests.ray.io

Seems that Flaky is reporting zero failures.

```
2023-06-16 12:07:41,873 INFO worker.py:1616 -- Started a local Ray instance.

 dashboard/modules/job/tests/test_cli_integration.py ⨯                                                                                    100% ██████████
================================================================ short test summary info ================================================================
FAILED test_cli_integration.py::TestRayAddress::test_empty_ray_address - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestRayAddress::test_valid_http_ray_address - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobSubmit::test_basic_submit - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobSubmit::test_submit_no_wait - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobSubmit::test_submit_with_logs_instant_job - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobSubmit::test_multiple_ray_init - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobSubmit::test_metadata - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestRuntimeEnv::test_bad_runtime_env - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobStop::test_basic_stop - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobStop::test_stop_no_wait - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobList::test_empty - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobList::test_list - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::TestJobDelete::test_basic_delete - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::test_quote_escaping - RuntimeError: Expected command to succeed, but got exit code: 1.
FAILED test_cli_integration.py::test_resources - RuntimeError: Expected command to succeed, but got exit code: 1.

Results (223.38s):
       3 passed
      15 failed
         - dashboard/modules/job/tests/test_cli_integration.py:109 TestRayAddress.test_empty_ray_address
         - dashboard/modules/job/tests/test_cli_integration.py:122 TestRayAddress.test_valid_http_ray_address
         - dashboard/modules/job/tests/test_cli_integration.py:129 TestJobSubmit.test_basic_submit
         - dashboard/modules/job/tests/test_cli_integration.py:136 TestJobSubmit.test_submit_no_wait
         - dashboard/modules/job/tests/test_cli_integration.py:143 TestJobSubmit.test_submit_with_logs_instant_job
         - dashboard/modules/job/tests/test_cli_integration.py:149 TestJobSubmit.test_multiple_ray_init
         - dashboard/modules/job/tests/test_cli_integration.py:157 TestJobSubmit.test_metadata
         - dashboard/modules/job/tests/test_cli_integration.py:167 TestRuntimeEnv.test_bad_runtime_env
         - dashboard/modules/job/tests/test_cli_integration.py:179 TestJobStop.test_basic_stop
         - dashboard/modules/job/tests/test_cli_integration.py:189 TestJobStop.test_stop_no_wait
         - dashboard/modules/job/tests/test_cli_integration.py:201 TestJobList.test_empty
         - dashboard/modules/job/tests/test_cli_integration.py:205 TestJobList.test_list
         - dashboard/modules/job/tests/test_cli_integration.py:220 TestJobDelete.test_basic_delete
         - dashboard/modules/job/tests/test_cli_integration.py:239 test_quote_escaping
         - dashboard/modules/job/tests/test_cli_integration.py:248 test_resources
 ```

I have a feeling that the first 28% of tests pass, then I noticed this error. I also tried running each test individually with pytest -vs, but had not much more luck. I think we are close, but unsure how to mitigate this. 
```
Traceback (most recent call last):
  File "~/.pyenv/versions/3.9.16/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
  File "~/aifactory/github/ray/python/ray/scripts/scripts.py", line 2474, in main
    return cli()
  File "~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "~aifactory/github/ray/python/ray/autoscaler/_private/cli_logger.py", line 856, in wrapper
    return f(*args, **kwargs)
  File "~/aifactory/github/ray/python/ray/scripts/scripts.py", line 742, in start
    raise ConnectionError(
ConnectionError: Ray is trying to start at 127.0.0.1:6379, but is already running at 127.0.0.1:6379. Please specify a different port using the `--port` flag of `ray start` command.
```

File: dashboard/modules/job/cli.py
Comment (architkulkarni) [inline dashboard/modules/job/cli.py:29]: I recommend removing this part, RAY_ADDRESS parsing currently happens elsewhere and this doesn't need to change in this PR as far as I can tell.

**Comment by architkulkarni** [comment]
The CI is running on Linux, I would guess that there's some flakiness when running locally on mac where Ray isn't fully getting shut down when it's expected to be.  You can manually run `ray stop` in case Ray is still running.  

But maybe it's quickest to merge master into your PR branch and focus on `test_cli.py`? Since that failed for this PR in CI, and passes for you locally on master.  I think you should be able to get it to pass locally.

**Comment by architkulkarni** [comment]
Screenshot 2023-06-16 at 12 43 16 PM
I couldn't reproduce the `pytest -vs test_cli_integration.py` failure on master. I'm on an intel mac

**Comment by peterghaddad** [comment]
@architkulkarni strange since I am also on an Intel MacBook. I will focus on 'test_cli.py' for this MR and look at ways to run the tests from a container. This would give developers a consistent environment when building and running Ray locally. However, I I'll continue to troubleshoot. 

Thank you for all the help on this!

File: doc/source/cluster/running-applications/job-submission/quickstart.rst
Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/quickstart.rst:84]: I think that would make sense!

**Comment by architkulkarni** [comment]
Possibly related failure: https://buildkite.com/ray-project/oss-ci-build-pr/builds/25962#0188d63c-b716-461c-8761-68b4fce3a831/1066-1555

Also, I think there are some unrelated failing tests, but I think they're fixed on master now. Can you try merging the latest master into your branch?

**Comment by peterghaddad** [comment]
Hey @architkulkarni I went ahead and merged master into my branch.

I had to modify quite a bit of the test file to more tests passing under: `pytest -vs test_cli.py::TestSubmit::test_job_id`

This is the error I am seeing now: 
```
result = <Result TypeError("submit() missing 1 required positional argument: 'verify'")>, exit_code = 0

    def check_exit_code(result, exit_code):
>       assert result.exit_code == exit_code, result.output
E       AssertionError: 
E       assert 1 == 0
E         +1
E         -0
```

Unsure how to fix these errors and don't want to modify existing tests. I was unsure if anything stuck out to you based on the changes I've pushed?

**Comment by architkulkarni** [comment]
The last argument to `submit` is `verify`.  In this PR, you added a new positional argument `headers`, increasing the total number of arguments by 1.  Somewhere in the code, a `submit` call hasn't been updated to include the new argument, so Python thinks the last argument `verify` is missing.  Are you able to reproduce it locally? Debugging should be much faster that way.

You might also need to update `mock_client_instance.submit_job.assert_called_with` calls in `test_cli.py` to expect that `submit_job` will be called with the `headers=None`.

**Comment by peterghaddad** [comment]
@architkulkarni Tests are working locally now and think we are in good shape! Can we get another look and potentially merge?

**Review by architkulkarni — COMMENTED** [review]
Just one last comment and then I think this is good to go! Thanks for your persistence

File: dashboard/modules/job/cli.py
Comment (peterghaddad) [reply dashboard/modules/job/cli.py]: Added! Let me know if this is what you were thinking.

**Review by architkulkarni — APPROVED** [review]
Looks good to me! Let me know when tests are passing (not counting flaky tests on https://flaky-tests.ray.io/) and I can merge it

**Comment by peterghaddad** [comment]
@architkulkarni Sounds good! Thanks. Tests look to be finished here or at least close to being finished. I pushed up some lint changes. Let me know if there is anything i'm missing! Thank you.

**Comment by peterghaddad** [comment]
@architkulkarni think it looks like we are good!

**Comment by architkulkarni** [comment]
@peterghaddad Can you check this test failure? https://buildkite.com/ray-project/oss-ci-build-pr/builds/30503#0189bd12-79a3-4ed3-8d9a-ff223b917b56/3503-3887

**Comment by peterghaddad** [comment]
@architkulkarni Seems like I am unable to check the test value returned from RaiseValueError because it triggers a subprocess run. Any workarounds you may know of? I simplified the test to still check for a test failure for the time being.

**Comment by architkulkarni** [comment]
That's fine, I don't know an easy way to do it off the top of my head.  The best way is probably to add an SDK test instead of a CLI test, feel free to do that in a follow-up if you like. Restarted the Python(large) tests because it looks like they didn't get triggered, but will merge once they finish

**Comment by peterghaddad** [comment]
@architkulkarni Sounds like a plan!

**Comment by architkulkarni** [comment]
Confirmed offline that https://buildkite.com/ray-project/oss-ci-build-pr/builds/30750#0189cffb-0bc7-4b52-a60b-6019015e8565/1738-1767 Python(large) failure is unrelated

However, this failure is related, https://buildkite.com/ray-project/oss-ci-build-pr/builds/30750#0189cffb-0bd9-4647-9ffb-0a67a933f6d7/1949-2381 @peterghaddad could you take a look?

File: doc/source/cluster/running-applications/job-submission/quickstart.rst
Comment (architkulkarni) [inline doc/source/cluster/running-applications/job-submission/quickstart.rst:82]: Might just need a newline after the code block? But best to check how it's done in other parts of the doc

Comment (peterghaddad) [reply doc/source/cluster/running-applications/job-submission/quickstart.rst]: Yep. That's it! Updated!

**Pull request merged**
